test/unit/math/opencl/kernel_generator/broadcast_test.cpp

```diff
@@ -0,0 +1,171 @@
+#ifdef STAN_OPENCL
+
+#include <stan/math/opencl/kernel_generator.hpp>
+#include <stan/math/opencl/matrix_cl.hpp>
+#include <stan/math/opencl/copy.hpp>
+#include <Eigen/Dense>
+#include <gtest/gtest.h>
+
+#define EXPECT_MATRIX_NEAR(A, B, DELTA) \
+  for (int i = 0; i < A.size(); i++)    \
+    EXPECT_NEAR(A(i), B(i), DELTA);
+
+TEST(MathMatrixCL, broadcast_errors) {
+  using Eigen::MatrixXd;
+  using stan::math::matrix_cl;
+
+  MatrixXd m(3, 3);
+  m << 1, 2.5, 3, 4, 5, 6.3, 7, -8, -9.5;
+  MatrixXd row(1, 3);
+  row << 10, 100, 1000;
+  MatrixXd col(3, 1);
+  col << 10, 100, 1000;
+  MatrixXd scal(1, 1);
+  scal << 10;
+
+  matrix_cl<double> m_cl(m);
+  matrix_cl<double> row_cl(row);
+  matrix_cl<double> col_cl(col);
+  matrix_cl<double> scal_cl(scal);
+
+  // extra parenthesis is needed for passing type name with comma in a macro
+  EXPECT_NO_THROW((stan::math::broadcast<false, false>(m_cl)));
+  EXPECT_THROW((stan::math::broadcast<true, false>(m_cl)),
+               std::invalid_argument);
+  EXPECT_THROW((stan::math::broadcast<false, true>(m_cl)),
+               std::invalid_argument);
+  EXPECT_THROW((stan::math::broadcast<true, true>(m_cl)),
+               std::invalid_argument);
+
+  EXPECT_NO_THROW((stan::math::broadcast<false, false>(row_cl)));
+  EXPECT_NO_THROW((stan::math::broadcast<true, false>(row_cl)));
+  EXPECT_THROW((stan::math::broadcast<false, true>(row_cl)),
+               std::invalid_argument);
+  EXPECT_THROW((stan::math::broadcast<true, true>(row_cl)),
+               std::invalid_argument);
+
+  EXPECT_NO_THROW((stan::math::broadcast<false, false>(col_cl)));
+  EXPECT_THROW((stan::math::broadcast<true, false>(col_cl)),
+               std::invalid_argument);
+  EXPECT_NO_THROW((stan::math::broadcast<false, true>(col_cl)));
+  EXPECT_THROW((stan::math::broadcast<true, true>(col_cl)),
+               std::invalid_argument);
+
+  EXPECT_NO_THROW((stan::math::broadcast<false, false>(scal_cl)));
+  EXPECT_NO_THROW((stan::math::broadcast<true, false>(scal_cl)));
+  EXPECT_NO_THROW((stan::math::broadcast<false, true>(scal_cl)));
+  EXPECT_NO_THROW((stan::math::broadcast<true, true>(scal_cl)));
+
+  EXPECT_NO_THROW((stan::math::broadcast<false, false>(scal_cl).eval()));
+  EXPECT_THROW((stan::math::broadcast<true, false>(scal_cl).eval()),
+               std::domain_error);
+  EXPECT_THROW((stan::math::broadcast<false, true>(scal_cl).eval()),
+               std::domain_error);
+  EXPECT_THROW((stan::math::broadcast<true, true>(scal_cl).eval()),
+               std::domain_error);
+}
+
+TEST(MathMatrixCL, broadcast_rows_test) {
+  using Eigen::MatrixXd;
+  using stan::math::matrix_cl;
+
+  MatrixXd m1(3, 3);
+  m1 << 1, 2.5, 3, 4, 5, 6.3, 7, -8, -9.5;
+  MatrixXd m2(1, 3);
+  m2 << 10, 100, 1000;
+
+  matrix_cl<double> m1_cl(m1);
+  matrix_cl<double> m2_cl(m2);
+
+  auto tmp = m1_cl + stan::math::broadcast<true, false>(m2_cl);
```

Comment:
Do you think we should follow Eigen's API here and have `rowwise()` and `colwise()` defined in the matrix_cl class?

Reply:
I think then this would turn into 
```cpp
auto tmp = m1_cl.rowwise() + m2_cl;
```

Reply:
While it makes the code look simpler in this case, I find it harder to understand. Also for more complex expressions it becomes less clear. For example:
```
matrix_cl m1(3,3), m2(3,3), v(3,1);
m1 + broadcast<true, false>(v) + m2;
```
becomes (I think):
```
m1.colwise() + v + m2.colwise();
```

If you want to avoud explicit template parameters I can instead do separate functions `rowwise_broadcast()` and `colwise_broadcast()`.

Reply:
Ah yes okay that makes a lot of sense! If you don't mind adding those I think they would be nice helper functions that would make the API a little cleaner

Reply:
done